src/hotspot/share/opto/phaseX.cpp

```diff
@@ -2269,6 +2269,49 @@ void PhasePeephole::print_statistics() {
 }
 #endif
 
+Node* PhaseLowering::transform(Node* n) {
+  // New node, make space for the node in the types table
+  _igvn->ensure_type_or_null(n);
+
+  // Try to find an existing version of the same node
+  Node* existing = _igvn->hash_find_insert(n);
+  if (existing != nullptr) {
+    // GVN hit, return the existing node
+    n->destruct(_igvn);
+    return existing;
+  }
+
+  // Record the type of the node in the types table
+  const Type* type = n->Value(_igvn);
+  _igvn->set_type(n, type);
+  n->raise_bottom_type(type);
+
+  return n;
+}
+
+
+void PhaseLowering::lower() {
+  // Worklist should be empty before lowering
+  _igvn->_worklist.ensure_empty();
+
+  C->identify_useful_nodes(_igvn->_worklist);
+
+  while(_igvn->_worklist.size() != 0) {
+    Node* n = _igvn->_worklist.pop();
+    Node* new_node = lower_node(n);
```

Comment:
I think we shouldn't run `Ideal` on the graph, because there is a chance that it could undo the lowering changes that we just did. This gives lowering more freedom to change the graph in different ways that would otherwise be undone by Ideal. We run `Value` on all of the transformed nodes mainly so the types table is accurate, so we can call upon the type of any node during lowering.

Reply:
> I think we shouldn't run `Ideal` on the graph, because there is a chance that it could undo the lowering changes that we just did. This gives lowering more freedom to change the graph in different ways that would otherwise be undone by Ideal. We run `Value` on all of the transformed nodes mainly so the types table is accurate, so we can call upon the type of any node during lowering.

We need to preserve lowering canonicalizations,  but lowered graph is susceptible to further idealizations, else we could have do lowering during final graph reshaping just before matching.